src/base/value_schema_v2.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "value_schema_v2.h"
+
+#include <dsn/utility/endians.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/c/api_utilities.h>
+#include <dsn/utility/smart_pointers.h>
+
+namespace pegasus {
+
+std::unique_ptr<value_field> value_schema_v2::extract_field(dsn::string_view value,
+                                                            value_field_type type)
+{
+    std::unique_ptr<value_field> field = nullptr;
+    switch (type) {
+    case value_field_type::EXPIRE_TIMESTAMP:
+        field = extract_timestamp(value);
+        break;
+    case value_field_type::TIME_TAG:
+        field = extract_time_tag(value);
+        break;
+    default:
+        dassert_f(false, "Unsupported field type: {}", type);
+    }
+    return field;
+}
+
+dsn::blob value_schema_v2::extract_user_data(std::string &&value)
+{
+    auto ret = dsn::blob::create_from_bytes(std::move(value));
+    return ret.range(sizeof(uint8_t) + sizeof(uint32_t) + sizeof(uint64_t));
```

Comment:
Can these values calculate in advance as a const value?

Reply:
How to calculate it in advance?

Reply:
I think @Smityz mean add a static const variable like `header_length = sizeof(uint8_t) + sizeof(uint32_t) + sizeof(uint64_t)` ?

Reply:
OK, I will refactor it in next pull request